File: java/file2bin/file2bin.java
//
//  Copy binary portion of file to another file
//  2021 (c) Dmitry Stefankov 
//

import java.io.*;
import java.util.List;
import java.nio.charset.Charset;
import java.nio.file.*;

public class file2bin {

  public static byte[] MyReadAllBytes(String filename) throws IOException {
	Path file = Paths.get(filename);
	return Files.readAllBytes(file);
  }

  public static void main(String[] args) throws IOException {
     //System.out.println("There are " + args.length + " arguments given.");
     //for(int i = 0; i < args.length; i++) 
    //    System.out.println("The argument #" + (i+1) + " is " + args[i] + " and is at index " + i);
     if (args.length < 4) {
       System.out.println("Copy binary portion of file to another file, version 1.0");
       System.out.println("Copyright (c) 2021, Dmitry Stefankov");
       System.out.println("Usage: file2bin infile outfile offset length");
       return;
     }
 
    String InFile = args[0];
     String OutFile = args[1];
     System.out.println("InFile: "+InFile);
     System.out.println("OutFile: "+OutFile);
     int offset = 0, length = 0;
     try {
       offset = Integer.parseInt(args[2]);
       System.out.println("Offset: "+offset);
       length = Integer.parseInt(args[3]);
       System.out.println("Length: "+length);
     }
     catch (NumberFormatException ex){
        ex.printStackTrace();
     }

     byte[] ReadBuffer = MyReadAllBytes(InFile);
     int InBufSize = ReadBuffer.length;
     System.out.println("InBufSize: "+InBufSize);

     byte[] NewBytes = new byte[length];
     for (int i = 0; i < NewBytes.length; i++) {
        NewBytes[i] = ReadBuffer[i+offset];
     }

     try (FileOutputStream stream = new FileOutputStream(OutFile)) {
       stream.write(NewBytes);
     }
  }

}

File: scala/file2bin/file2bin.scala
// Scala Program on command line argument
import java.io.File
import java.io.PrintWriter
import scala.sys.process._
import scala.language.postfixOps
import scala.io.Source._
import java.nio.charset.StandardCharsets
import java.nio.file.{Files, Path}
//import java.util.Scanner
//import java.io.File

object File2bin
{
    // Main method
    def main(args: Array[String])
    {
       if (args.length == 0)
       {
         println("Copy binary portion of file to another file, version 1.0");
         println("Copyright (c) 2025, Dmitry Stefankov");
         println("Usage: file2bin infile outfile offset length");
         sys.exit(1);
       }

        // You pass any thing at runtime 
        // that will be print on the console
        var count : Int = 1;
        var infile : String = ""
        var outfile : String = ""
        var filelen : Int = 0
        var fileofs : Int = 0
        for(arg<-args)
        {
            println("Arg " + count + ": " + arg);
            if (count == 1) { infile = arg; println(infile) }
            if (count == 2) { outfile = arg; println(outfile) }
            if (count == 3) { fileofs = arg.toInt; println(fileofs) }
            if (count == 4) { filelen = arg.toInt; println(filelen) }
            count += 1
        }

        // --- read input file
        val path = Path.of(infile)
        //var content: String = os.read(path)
        //val lines = fromFile(path).getLines
        val source = scala.io.Source.fromFile(infile)
        val lines = try source.mkString finally source.close()
        //println(lines)
        val sub1 = lines.substring(fileofs,fileofs+filelen)

        // Creating a file 
        val file_Object = new File(outfile)
        // Passing reference of file to the printwriter     
        val print_Writer = new PrintWriter(file_Object)
        // Writing to the file       
        //print_Writer.write(lines)
        print_Writer.write(sub1)
        // Closing printwriter    
        print_Writer.close()

    }
}

